File: v1/HLD/Capstones/code/RateLimiter.java
public class RateLimiter {
    static final double CAPACITY = 5.0;
    static final double REFILL_RATE_PER_SEC = 2.0;

    static class TokenBucket {
        double capacity;
        double refillRate;
        double tokens;
        long lastRefillNanos;

        TokenBucket(double capacity, double refillRate) {
            this.capacity = capacity;
            this.refillRate = refillRate;
            this.tokens = capacity;
            this.lastRefillNanos = System.nanoTime();
        }

        void refill() {
            long now = System.nanoTime();
            double elapsedSeconds = (now - lastRefillNanos) / 1_000_000_000.0;
            double added = elapsedSeconds * refillRate;
            if (added > 0) {
                tokens = Math.min(capacity, tokens + added);
                lastRefillNanos = now;
            }
        }

        boolean allowRequest() {
            refill();
            if (tokens >= 1.0) {
                tokens -= 1.0;
                return true;
            }
            return false;
        }
    }

    public static void main(String[] args) throws InterruptedException {
        TokenBucket bucket = new TokenBucket(CAPACITY, REFILL_RATE_PER_SEC);
        long start = System.nanoTime();

        System.out.println("Burst of 8 requests with no delay (capacity 5):");
        for (int i = 0; i < 8; i++) {
            boolean allowed = bucket.allowRequest();
            double elapsedMs = (System.nanoTime() - start) / 1_000_000.0;
            System.out.printf("  request %d at %.0f ms: %s%n", i + 1, elapsedMs, allowed ? "allowed" : "REJECTED (rate limited)");
        }

        System.out.println("Waiting 1.5s for tokens to refill...");
        Thread.sleep(1500);

        System.out.println("3 more requests after waiting:");
        for (int i = 0; i < 3; i++) {
            boolean allowed = bucket.allowRequest();
            double elapsedMs = (System.nanoTime() - start) / 1_000_000.0;
            System.out.printf("  request %d at %.0f ms: %s%n", i + 1, elapsedMs, allowed ? "allowed" : "REJECTED (rate limited)");
        }
    }
}

File: v1/HLD/Capstones/code/rate_limiter.js
const CAPACITY = 5;
const REFILL_RATE_PER_SEC = 2;

class TokenBucket {
  constructor(capacity, refillRate) {
    this.capacity = capacity;
    this.refillRate = refillRate;
    this.tokens = capacity;
    this.lastRefill = process.hrtime.bigint();
  }

  refill() {
    const now = process.hrtime.bigint();
    const elapsedSeconds = Number(now - this.lastRefill) / 1e9;
    const added = elapsedSeconds * this.refillRate;
    if (added > 0) {
      this.tokens = Math.min(this.capacity, this.tokens + added);
      this.lastRefill = now;
    }
  }

  allowRequest() {
    this.refill();
    if (this.tokens >= 1) {
      this.tokens -= 1;
      return true;
    }
    return false;
  }
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

(async () => {
  const bucket = new TokenBucket(CAPACITY, REFILL_RATE_PER_SEC);
  const start = process.hrtime.bigint();

  console.log("Burst of 8 requests with no delay (capacity 5):");
  for (let i = 0; i < 8; i++) {
    const allowed = bucket.allowRequest();
    const elapsedMs = Number(process.hrtime.bigint() - start) / 1e6;
    console.log(`  request ${i + 1} at ${elapsedMs.toFixed(0)} ms: ${allowed ? "allowed" : "REJECTED (rate limited)"}`);
  }

  console.log("Waiting 1.5s for tokens to refill...");
  await sleep(1500);

  console.log("3 more requests after waiting:");
  for (let i = 0; i < 3; i++) {
    const allowed = bucket.allowRequest();
    const elapsedMs = Number(process.hrtime.bigint() - start) / 1e6;
    console.log(`  request ${i + 1} at ${elapsedMs.toFixed(0)} ms: ${allowed ? "allowed" : "REJECTED (rate limited)"}`);
  }
})();
